In [1]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [2]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [3]:
# Generate summary statistics by column
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [4]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

The above chart displays price change percent across different time periods for each type of cryptocurrency.
We can see that the currencies Ethlend, Celcius-Degree-Token and Theta-Token are the most volatile.

---

### Prepare the Data

In [5]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
column_names = ["price_change_percentage_24h","price_change_percentage_7d","price_change_percentage_14d","price_change_percentage_30d","price_change_percentage_60d","price_change_percentage_200d","price_change_percentage_1y"]
market_data_scaled = StandardScaler().fit_transform(df_market_data[column_names])

# Check the first 2 rows of scaled data: 
market_data_scaled[0:2]

array([[ 0.50852937,  0.49319307,  0.77220043,  0.23545963, -0.0674951 ,
        -0.35595348, -0.25163688],
       [ 0.18544589,  0.93444504,  0.55869212, -0.05434093, -0.27348273,
        -0.11575947, -0.19935211]])

In [6]:
# Create a DataFrame with the scaled data
market_data_scaled_df = pd.DataFrame(
        market_data_scaled,
        columns = column_names
        )
#market_data_scaled_df.head()

# Copy the crypto names from the original DataFrame
market_data_scaled_df["coin_id"] = df_market_data.index

# Set the coin_id column as index
market_data_scaled_df = market_data_scaled_df.set_index("coin_id")
#market_data_scaled_df.index = market_data_scaled_df.index.get_level_values(0)

# Display the scaled DataFrame
market_data_scaled_df.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Scaled DataFrame.

In [7]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1, 11))

In [8]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list

for i in k:
    k_model = KMeans(n_clusters=i, random_state=0)
    k_model.fit(market_data_scaled_df)
    inertia.append(k_model.inertia_)

In [9]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {"k": k, "inertia": inertia}

# Create the DataFrame from the elbow data
df_elbow = pd.DataFrame(elbow_data)

# Review the DataFrame
df_elbow.head()

k     inertia
0  1  287.000000
1  2  198.571818
2  3  145.635748
3  4   79.022435
4  5   66.217780

In [42]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
elbow = df_elbow.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=k
)
elbow

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** Based on the chart, the "elbow" of the curve happens at cluster 4, which is where the inertia of the curve starts to plateau. It is worth also testing 5 clusters with other methods, such as the Silhouette Method or Calinski Harabasz Score to confirm that 4 is the optimal number of clusters.

---

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

In [11]:
# Initialize the K-Means model using the best value for k
# Random state is used for repeatability
model = KMeans(n_clusters=4, random_state=0)

In [12]:
# Fit the K-Means model using the scaled DataFrame
model.fit(market_data_scaled_df)

KMeans(n_clusters=4, random_state=0)

In [13]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
k_4 = model.predict(market_data_scaled_df)

# Print the resulting array of cluster values.
k_4

array([2, 2, 0, 0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 2, 0, 2, 0, 0, 2, 0, 0, 2,
       0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 1, 2, 0, 0, 3, 0, 0, 0, 0],
      dtype=int32)

In [14]:
# Create a copy of the scaled DataFrame
market_data_scaled_predictions = market_data_scaled_df.copy()
#market_data_scaled_predictions.head()

In [15]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters
market_data_scaled_predictions["cluster_prediction"] = k_4
market_data_scaled_predictions.columns = market_data_scaled_predictions.columns.get_level_values(0)

# Display the copy of the scaled DataFrame
#market_data_scaled_predictions.columns = market_data_scaled_predictions.columns.astype(str)
market_data_scaled_predictions.tail()


price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
celsius-degree-token                     1.045530                   -0.618328   
ontology                                -0.409044                   -0.906963   
ftx-token                                0.414711                    0.414044   
true-usd                                 0.078038                   -0.687745   
digibyte                                 1.217453                   -0.607714   

                      price_change_percentage_14d  \
coin_id                                             
celsius-degree-token                     2.907054   
ontology                                -1.298986   
ftx-token                               -0.047386   
true-usd                                -0.009191   
digibyte                                -0.907066   

                      price_change_percentage_30d  \
coin_id                                             
celsius-degree-token                     5.351455   
ontology                                -1.393153   
ftx-token                               -0.465380   
true-usd                                -0.058214   
digibyte                                 0.449939   

                      price_change_percentage_60d  \
coin_id                                             
celsius-degree-token                     4.769913   
ontology                                -0.696937   
ftx-token                                0.128185   
true-usd                                 0.007388   
digibyte                                -0.662530   

                      price_change_percentage_200d  \
coin_id                                              
celsius-degree-token                      3.148875   
ontology                                 -0.422835   
ftx-token                                -0.416502   
true-usd                                 -0.550441   
digibyte                                  0.572367   

                      price_change_percentage_1y  cluster_prediction  
coin_id                                                               
celsius-degree-token                    1.348488                   3  
ontology                               -0.292344                   0  
ftx-token                              -0.145469                   0  
true-usd                               -0.281747                   0  
digibyte                               -0.132482                   0

In [16]:
print(market_data_scaled_predictions.columns)

Index(['price_change_percentage_24h', 'price_change_percentage_7d',
       'price_change_percentage_14d', 'price_change_percentage_30d',
       'price_change_percentage_60d', 'price_change_percentage_200d',
       'price_change_percentage_1y', 'cluster_prediction'],
      dtype='object')


In [39]:
# Create a scatter plot using hvPlot by setting
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.

scaled_plot = market_data_scaled_predictions.hvplot.scatter(
    x = "price_change_percentage_24h",
    y = "price_change_percentage_7d",
    by = "cluster_prediction",
    hover_cols=["coin_id"],
    title = "Cryptocurrencies Clusters | Scaled Data"
)

scaled_plot

:NdOverlay   [cluster_prediction]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

### Optimize Clusters with Principal Component Analysis.

In [18]:
# Import the PCA Module:
from sklearn.decomposition import PCA

# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components = 3)

In [19]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame
# down to three principal components.
market_data_pca = pca.fit_transform(market_data_scaled_df)

# View the scaled PCA data
market_data_pca[:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [20]:
# Retrieve the explained variance to determine how much information
# can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.3719856 , 0.34700813, 0.17603793])

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** The total explained variance of the three principal components is .895, meaning that approximately 90% of the data in the dataframe is explained by the three main principal components. PCA 1 explains 37.2% of the data, PCA 2 explains 34.7%, and PCA 3 explains 17.6%.

In [21]:
# Create a new DataFrame with the PCA data.
market_data_pca_df = pd.DataFrame(
    market_data_pca,
    columns=["PCA1","PCA2","PCA3"]
)

# Copy the crypto names from the original scaled DataFrame
market_data_pca_df["coin_id"] = market_data_scaled_df.index

# Set the coin_id column as index
market_data_pca_df = market_data_pca_df.set_index("coin_id")

# Display the scaled PCA DataFrame
market_data_pca_df.head()

PCA1      PCA2      PCA3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the Scaled PCA DataFrame

Similar to our initia K-Means Cluster Analysis, we repeat the elbow method and plotting of the clusters, but instead focus on the 3 PCA values we derived in the above code

In [22]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1, 11))

In [23]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list

for i in k:
    k_model = KMeans(n_clusters=i, random_state=0)
    k_model.fit(market_data_pca_df)
    inertia.append(k_model.inertia_)

In [24]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_pca = {"k": k, "inertia": inertia}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow_pca = pd.DataFrame(elbow_data_pca)
df_elbow_pca.head()

k     inertia
0  1  256.874086
1  2  168.811896
2  3  115.918077
3  4   49.665497
4  5   41.357314

In [41]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
elbow_pca = df_elbow_pca.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=k
)
elbow_pca

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** The best value for "k" using the PCA data is 4, as that is where the intertia values begin to plateau/decellerate. 


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** Using the intial elbow method on the scaled K-Clusters, we also determined that 4 was the optimal cluster number. However, the PCA method made the decelleration of the inertia values more obvious than the initial elbow curve, where the cluster number could be argued as either 4 or 5. 

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [26]:
# Initialize the K-Means model using the best value for k
model_pca = KMeans(n_clusters=4, random_state=0)

In [27]:
# Fit the K-Means model using the PCA data
model.fit(market_data_pca_df)

KMeans(n_clusters=4, random_state=0)

In [28]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
k_4_pca = model.predict(market_data_pca_df)

# Print the resulting array of cluster values.
k_4_pca

array([2, 2, 0, 0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 2, 0, 2, 0, 0, 2, 0, 0, 2,
       0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 1, 2, 0, 0, 3, 0, 0, 0, 0],
      dtype=int32)

In [29]:
# Create a copy of the scaled PCA DataFrame
market_data_pca_predictions = market_data_pca_df.copy()

# Add a new column to the copy of the PCA DataFrame with the predicted clusters
market_data_pca_predictions["pca_cluster_prediction"] = k_4_pca

# Display the copy of the scaled PCA DataFrame
market_data_pca_predictions.head()

PCA1      PCA2      PCA3  pca_cluster_prediction
coin_id                                                           
bitcoin      -0.600667  0.842760  0.461595                       2
ethereum     -0.458261  0.458466  0.952877                       2
tether       -0.433070 -0.168126 -0.641752                       0
ripple       -0.471835 -0.222660 -0.479053                       0
bitcoin-cash -1.157800  2.041209  1.859715                       2

In [37]:
# Create a scatter plot using hvPlot by setting
# `x="PC1"` and `y="PC2"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.

clusters_pca_plot = market_data_pca_predictions.hvplot.scatter(
    x="PCA1",
    y="PCA2",
    by="pca_cluster_prediction",
    hover_cols=["coin_id"],
    title = "Cryptocurrencies Clusters | PCA Data"
)

clusters_pca_plot

:NdOverlay   [pca_cluster_prediction]
   :Scatter   [PCA1]   (PCA2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [43]:
# Composite plot to contrast the Elbow curves
elbow + elbow_pca

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [40]:
# Composite plot to contrast the clusters
scaled_plot + clusters_pca_plot

:Layout
   .NdOverlay.I  :NdOverlay   [cluster_prediction]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [pca_cluster_prediction]
      :Scatter   [PCA1]   (PCA2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** Using fewer features allowed us to better isolate the differences amongst the four clusters. This is especially prevalent for cluster 3, which initially was indecipherable amongst cluster 0 data, while now, we can identify it as having a high values across both PCA 1 and PCA 2. The individual datapoints in each cluster are also now closer together & are closer to their centroids, while in the initial scaled chart the datapoints were more spread out. We can get an overall more cohesive & streamlined story using the PCA method, allowing us to develop cluster summaries (below):

  * Cluster 0: These datapoints have both a low PCA1 value and a low PCA2 value

  * Cluster 1: This datapoint sits in its own cluster, as it has a significantly high PCA 1 value but significantly low PCA2 value

  * Cluster 2: These datapoints also have low PCA1 values (similar to cluster 0) but have higher PCA2 values

  * Cluster 3: This datapoint sits in its own cluster, as it has both a high PCA1 value and a high PCA2 value